### ISSUE TITLE
how can I generate the heatmap-liked pictures?

### ISSUE THREAD

**Issue by omtrix** [zengqunzhao/MA-Net#6]
in your article,I notice you use the heatmap-liked pictures to show the where the net is focusing.
it's cool, however, when I tried to use pytorch-grad-cam to do this,I found out that the net has 2 fc layers(different from the nomal net like resnet50) and i failed to generate the pictures. 
Can you tell me your method?

**Comment by zengqunzhao** [comment]
> in your article,I notice you use the heatmap-liked pictures to show the where the net is focusing. it's cool, however, when I tried to use pytorch-grad-cam to do this,I found out that the net has 2 fc layers(different from the nomal net like resnet50) and i failed to generate the pictures. Can you tell me your method?

Some main codes, it may helpful:
`for i, (images, target) in enumerate(val_loader):

    label = target.cpu()
    label = label.reshape(args.batch_size, 1)

    one_hot = torch.zeros(args.batch_size, 7).scatter_(1, label, 1)
    one_hot = one_hot.unsqueeze(2).expand(args.batch_size, 7, 14 * 14)

    # compute output
    output1, output2, feature1, feature2 = model(images)

    feature1 = feature1.cpu()
    bz, nc, h, w = feature1.shape
    feature1 = feature1.reshape((bz, nc, h * w))
    feature1 = feature1.permute(0, 2, 1)
    feature1 = feature1.reshape(bz * h * w, nc)

    for name, value in model.named_parameters():
        if name == "module.fc_1.weight":
            weight_softmax_1 = value
        if name == "module.fc_2.weight":
            weight_softmax_2 = value

    weight_softmax_1 = weight_softmax_1.cpu()

    weight_softmax_1 = weight_softmax_1.permute(1, 0)

    f_out_1 = feature1.mm(weight_softmax_1)
    f_out_1 = f_out_1.reshape(bz, h * w, 7)
    f_out_1 = f_out_1.permute(0, 2, 1)

    f_1 = one_hot.mul(f_out_1)
    f_1 = torch.sum(f_1, 1, keepdim=False)
    f_1 = f_1.reshape(bz, h, w)
    f_1 = f_1.detach().numpy()`

**Comment by zengqunzhao** [comment]
> in your article,I notice you use the heatmap-liked pictures to show the where the net is focusing. it's cool, however, when I tried to use pytorch-grad-cam to do this,I found out that the net has 2 fc layers(different from the nomal net like resnet50) and i failed to generate the pictures. Can you tell me your method?


Some main codes, it may helpful:
  for i, (images, target) in enumerate(val_loader):
      label = target.cpu()
      label = label.reshape(args.batch_size, 1)
      one_hot = torch.zeros(args.batch_size, 7).scatter_(1, label, 1)
      one_hot = one_hot.unsqueeze(2).expand(args.batch_size, 7, 14 * 14)
  
      # compute output
      output1, output2, feature1, feature2 = model(images)
  
      feature1 = feature1.cpu()
      bz, nc, h, w = feature1.shape
      feature1 = feature1.reshape((bz, nc, h * w))
      feature1 = feature1.permute(0, 2, 1)
      feature1 = feature1.reshape(bz * h * w, nc)
  
      for name, value in model.named_parameters():
          if name == "module.fc_1.weight":
              weight_softmax_1 = value
          if name == "module.fc_2.weight":
              weight_softmax_2 = value
  
      weight_softmax_1 = weight_softmax_1.cpu()
      weight_softmax_1 = weight_softmax_1.permute(1, 0)
  
      f_out_1 = feature1.mm(weight_softmax_1)
      f_out_1 = f_out_1.reshape(bz, h * w, 7)
      f_out_1 = f_out_1.permute(0, 2, 1)
  
      f_1 = one_hot.mul(f_out_1)
      f_1 = torch.sum(f_1, 1, keepdim=False)
      f_1 = f_1.reshape(bz, h, w)
      f_1 = f_1.detach().numpy()

**Comment by zengqunzhao** [comment]
> in your article,I notice you use the heatmap-liked pictures to show the where the net is focusing. it's cool, however, when I tried to use pytorch-grad-cam to do this,I found out that the net has 2 fc layers(different from the nomal net like resnet50) and i failed to generate the pictures. Can you tell me your method?

Some main codes, it may help:
```
for i, (images, target) in enumerate(val_loader):

    label = target.cpu()
    label = label.reshape(args.batch_size, 1)

    one_hot = torch.zeros(args.batch_size, 7).scatter_(1, label, 1)
    one_hot = one_hot.unsqueeze(2).expand(args.batch_size, 7, 14 * 14)

    # compute output
    output1, output2, feature1, feature2 = model(images)

    feature1 = feature1.cpu()
    bz, nc, h, w = feature1.shape
    feature1 = feature1.reshape((bz, nc, h * w))
    feature1 = feature1.permute(0, 2, 1)
    feature1 = feature1.reshape(bz * h * w, nc)

    for name, value in model.named_parameters():
        if name == "module.fc_1.weight":
            weight_softmax_1 = value
        if name == "module.fc_2.weight":
            weight_softmax_2 = value

    weight_softmax_1 = weight_softmax_1.cpu()
    weight_softmax_1 = weight_softmax_1.permute(1, 0)

    f_out_1 = feature1.mm(weight_softmax_1)
    f_out_1 = f_out_1.reshape(bz, h * w, 7)
    f_out_1 = f_out_1.permute(0, 2, 1)

    f_1 = one_hot.mul(f_out_1)
    f_1 = torch.sum(f_1, 1, keepdim=False)
    f_1 = f_1.reshape(bz, h, w)
    f_1 = f_1.detach().numpy()
```
